internal/dialect/dialectquery/clickhouse.go

```diff
@@ -11,10 +23,23 @@ func (c *Clickhouse) CreateTable(tableName string) string {
 		version_id Int64,
 		is_applied UInt8,
 		date Date default now(),
-		tstamp DateTime default now()
+		tstamp DateTime64(9, 'UTC') default now64(9, 'UTC')
 	  )
-	  ENGINE = MergeTree()
-		ORDER BY (date)`
+	  ENGINE = KeeperMap('/goose_version')
```

Comment:
I've been coming back to this PR on/off, trying to think through the implementation from the perspective of a "general purpose" migration tool. On the surface, everything looks reasonable with the exception of changing the engine.

For context, another PR https://github.com/pressly/goose/pull/530 came in requesting another engine.

Should the default engine be `MergeTree` but allow this to be configurable?

Maybe we could put our heads together on how to solve this so the tool can be useful for all or most ClickHouse users? cc @iamluc @arunk-s 

I think PR https://github.com/pressly/goose/pull/530 is on the right track to setting a custom engine, and afaics this would satisfy:

1. default engine: `MergeTree()`
2. `ReplicatedMergeTree()` 
3. `KeeperMap()`
4. ... etc

Other questions, does the `ORDER BY` or `PARTITION BY` need to be configurable? Hopefully not so we can make this as generic as possible.

Last thought, leaving `ENGINE = %s` to the user _may_ be a footgun and I wish we could do more upfront validation, but then we have to account for all the possible ClickHouse engine types and this seems like a lot of work to maintain.

Maybe there's a world where the user is allowed to supply a template for the table and as long as it contains:

- `version_id`
- `is_applied` (legacy field, deprecated)
- `tstamp`

Not sure I like this idea, since goose should own the table it creates to ensure the identifiers match up across all implementations.

Reply:
Hi,
I had a quick look at the PR and it looks good to me (and could replace my own PR https://github.com/pressly/goose/pull/530)

My only concern is about the engine. I'm not a ClickHouse expert and didn't know about the `KeeperMap` engine, but it seems it must be enabled (https://clickhouse.com/docs/en/engines/table-engines/special/keeper-map) which could be not possible with hosted services.
I suggest to add the engine as a new option.

Reply:
Having an option to toggle underlying table engine can definitely work!

> Other questions, does the ORDER BY or PARTITION BY need to be configurable? Hopefully not so we can make this as generic as possible.

I think these settings are affected by the choice of Engine so yeah they'll need to be configurable if the custom engine is allowed.

I can understand the concern with hosted services to not have `KeeperMap` though I doubt that platforms like ClickHouse Cloud disallows creating tables with this table engine. 

> Last thought, leaving ENGINE = %s to the user may be a footgun and I wish we could do more upfront validation, but then we have to account for all the possible ClickHouse engine types and this seems like a lot of work to maintain.

Definitely, convention over configuration is generally desirable however in this particular case I'm afraid the abstractions will leak a bit in order to properly support the underlying database. 

The library does quite well supporting different databases but that does come at a cost I'm afraid.

I'll be happy to lend a hand with ClickHouse specific topics though it can take a bit of time for me to catch up.